### PR TITLE
Make .async-search-* a restricted namespace

### PR DESCRIPTION
**Pull request by albertzaharovits**
ES Security hides certain indices, which should not be accessed by non-superusers.
This PR hides the `.async-search-*` namespace, the same way `.security` is hidden. The value of the namespace is hard-coded.

I have qualms about ES Security hiding an index which is not its own. Ideally, other plugins should have the capability to register such indices, but I don't see an easy way to achieve this. 

I plan to raise a follow-up that adds a new origin and system user that makes use of the new restricted index namespace. Also, the ILM origin should be adjusted (also in a follow-up) to allow ILM to work over this namespace.

PS this can also be targeted to a feature branch if this is more appropriate.

CC @jimczi 
Relates https://github.com/elastic/elasticsearch/pull/49931

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/regex/Regex.java [hunk 1]
```diff
@@ -44,6 +44,13 @@ public static boolean isSimpleMatchPattern(String str) {
         return str.indexOf('*') != -1;
     }
 
+    /**
+     * Is the str a lucene type of pattern
+     */
+    public static boolean isLuceneRegex(String str) {
+        return str.length() > 1 && str.charAt(0) == '/' && str.charAt(str.length() - 1) == '/';
```
Comment (tvernum) [inline server/src/main/java/org/elasticsearch/common/regex/Regex.java]: I don't think this "regex" style is used outside of security, so I don't think we need to add this method to `server/`
Maybe `Automatons` would be better?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/index/RestrictedIndicesNames.java [hunk 2]
```diff
@@ -21,10 +23,21 @@
     public static final String INTERNAL_SECURITY_TOKENS_INDEX_7 = ".security-tokens-7";
     public static final String SECURITY_TOKENS_ALIAS = ".security-tokens";
 
+    // public for tests
+    public static final String ASYNC_SEARCH_PREFIX = ".async-search-";
+    private static final CharacterRunAutomaton ASYNC_SEARCH_AUTOMATON =
+            new CharacterRunAutomaton(Automatons.patterns(ASYNC_SEARCH_PREFIX + "*"));
```
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/index/RestrictedIndicesNames.java]: An automaton seems like overkill for a prefix, could we just use `startsWith` in `isRestricted` ?

File: server/src/main/java/org/elasticsearch/common/regex/Regex.java (hunk 1 above)
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/common/regex/Regex.java]: Yes, I think it's better, thanks.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/index/RestrictedIndicesNames.java (hunk 2 above)
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/index/RestrictedIndicesNames.java]: Sounds fair, I made the change as suggested.

**Comment by albertzaharovits** [comment]
@tvernum  I've addressed the review comments, please take another look.

**Review by tvernum — APPROVED** [review]
LGTM

**Pull request merged**